runtime/ops/fs.rs

```diff
@@ -820,58 +796,110 @@ async fn op_remove_async(
   .unwrap()
 }
 
-#[derive(Deserialize)]
-#[serde(rename_all = "camelCase")]
-pub struct CopyFileArgs {
-  from: String,
-  to: String,
-}
-
 #[op]
 fn op_copy_file_sync(
   state: &mut OpState,
-  args: CopyFileArgs,
+  from: String,
+  to: String,
 ) -> Result<(), AnyError> {
-  let from = PathBuf::from(&args.from);
-  let to = PathBuf::from(&args.to);
+  let from_path = PathBuf::from(&from);
+  let to_path = PathBuf::from(&to);
 
   let permissions = state.borrow_mut::<Permissions>();
-  permissions.read.check(&from)?;
-  permissions.write.check(&to)?;
+  permissions.read.check(&from_path)?;
+  permissions.write.check(&to_path)?;
 
-  debug!("op_copy_file_sync {} {}", from.display(), to.display());
   // On *nix, Rust reports non-existent `from` as ErrorKind::InvalidInput
   // See https://github.com/rust-lang/rust/issues/54800
   // Once the issue is resolved, we should remove this workaround.
-  if cfg!(unix) && !from.is_file() {
+  if cfg!(unix) && !from_path.is_file() {
     return Err(custom_error(
       "NotFound",
       format!(
         "File not found, copy '{}' -> '{}'",
-        from.display(),
-        to.display()
+        from_path.display(),
+        to_path.display()
       ),
     ));
   }
 
   let err_mapper = |err: Error| {
     Error::new(
       err.kind(),
-      format!("{}, copy '{}' -> '{}'", err, from.display(), to.display()),
+      format!(
+        "{}, copy '{}' -> '{}'",
+        err,
+        from_path.display(),
+        to_path.display()
+      ),
     )
   };
+
+  #[cfg(target_os = "macos")]
+  {
+    use libc::chmod;
+    use libc::clonefile;
+    use libc::stat;
+    use libc::unlink;
+    use std::ffi::CString;
+    use std::io::Read;
+
+    let from = CString::new(from).unwrap();
+    let to = CString::new(to).unwrap();
+
+    // SAFETY: `from` and `to` are valid C strings.
+    // std::fs::copy does open() + fcopyfile() on macOS. We try to use
+    // clonefile() instead, which is more efficient.
+    unsafe {
```

Comment:
i've just read the clonefile man page. is it correct to use clonefile here as the new file will have different semantics (COW) than if we had done a regular copy and caller won't be aware of this. seems to me like clonefile is something caller should have to explicitly ask for, no?

Reply:
maybe i am misunderstanding? so, if the underlying FS supports COW then the result of clonefile will be the same as doing a regular copy on that FS? if so, then that sounds fine.

Reply:
yes - that makes sense. APFS on mac seems to support COW out of the box. i just wonder if OS reports disk usage etc. differently when you use clonefile and maybe this causing confusion? whenever the new file is changed, a copy will be made by the OS anyway so effect should be the same.